netlify/remix-template
# fix: support npm run dev (remix dev) and Netlify CLI

## Description

This fix when the `Functions Template` is selected, allows a developer to use `npm run dev` as they would expect, i.e. run `remix dev` and at the same time allow for use of the Netlify CLI (remix watch).

## Related Tickets & Documents



- Related Issue #83
- Closes #

## QA Instructions, Screenshots, Recordings

1. Ensure the [Netlify CLI is installed and up to date](https://docs.netlify.com/cli/get-started/).
3. There are two test scenarios:
    * Create a Remix app using TypeScript for Netlify Functions
    * Create a Remix app using JavaScript for Netlify Functions
    * 
4. For each scenario, it always starts by running the following from the root folder in a shell.

```bash
npx create-remix --template ./
```

5. The CLI will prompt for a location for the project.

```bash
❯ npx create-remix --template ./
? Where would you like to create your app? (./my-remix-app)
```

Ensure that the location is outside of the  Netlify Remix template folder, e.g. `../my-remix-app`

6. The Remix CLI will prompt you to choose TypeScript or JavaScript. Choose accordingly for the given test scenarios mentioned above.

```bash
? TypeScript or JavaScript? (Use arrow keys)
❯ TypeScript 
  JavaScript
```

7. The Remix CLI will prompt you to install packages, press the <kbd>ENTER</kbd> to accept the default answer of Yes

```bash
? Do you want me to run `npm install`? (Y/n)
```

The npm packages will install.

8. Some files will be manipulated if JavaScript is selected. The Remix CLI will output something similar to this.

```
⠹ Creating your app…⠋ Migrating template to JavaScript…Processing 3 files... 
Spawning 3 workers...
Sending 1 files to free worker...
Sending 1 files to free worker...
Sending 1 files to free worker...
⠴ Migrating template to JavaScript… SKIP /Users/nicktaylor/dev/remix-tests/edgy-ts/globals.d.ts
⠦ Migrating template to JavaScript… OKK /Users/nicktaylor/dev/remix-tests/edgy-ts/remix.init/root.tsx
 OKK /Users/nicktaylor/dev/remix-tests/edgy-ts/remix.init/entry.server.tsx
All done. 
Results: 
0 errors
0 unmodified
1 skipped
2 ok
Time elapsed: 0.428seconds
```

9. The Remix CLI extended with our questions will prompt to create the app with Netlify Functions or Netlify Edge Functions.

```bash
💿 Running remix.init script
? Run your Remix site with: (Use arrow keys)
❯ Netlify Functions - Choose this for stable support for production sites 
  Netlify Edge Functions (beta) - Try this for improved performance on non-critical sites
```

10 Select *Netlify Functions* by pressing the ENTER key.
10. Change to the directory of the project you just created, e.g. `../my-remix-app`
13. The `netlify.toml` for the project should be the same as the one in this PR.
14. Run `npm install`

### Test `remix dev`

1. Run

```bash
npm run dev
```

2. The Remix dev server starts

```bash
➜ npm run dev

> dev
> remix dev

Remix App Server started at http://localhost:3000 (http://192.168.86.78:3000)
This is an example of how to set caching headers for a route, feel free to change the value of 60 seconds or remove the header
GET / 200 - - 38.115 ms
This is an example of how to set caching headers for a route, feel free to change the value of 60 seconds or remove the header
GET / 200 - - 2.539 ms
```
3. Navigate to http//:localhost:3000 to see the page load.

![CleanShot 2023-02-27 at 16 52 03](https://user-images.githubusercontent.com/833231/221694196-3490b47b-1581-4350-be52-2a0b4d9731d6.png)

5. Make some changes to the text in app/routes/index.tsx (or jsx)
6. The page reloads with the changes.

### Test with Netlify CLI `ntl dev`

1. Run `ntl dev`
2. The Remix dev server starts up.
3. Navigate to `https://localhost:8888` and the page loads.
4. Make some changes to the text in app/routes/index.tsx (or jsx)
5. The page reloads with the changes.

![image](https://user-images.githubusercontent.com/833231/219900482-efe32bad-fc63-429b-a4df-874601e3e78a.png)
For us to review and ship your PR efficiently, please perform the following steps:

### Test with Netlify CLI `ntl serve`

1. Run `npm run start`
2. `ntl serve runs` (not the Remix dev server)
3. The `_custom_redirects` file gets copied to `public/_redirects`
4. Navigate to `https://localhost:8888` and the page loads.
 
![image](https://user-images.githubusercontent.com/833231/219900482-efe32bad-fc63-429b-a4df-874601e3e78a.png)
For us to review and ship your PR efficiently, please perform the following steps:

      can discuss the changes and get feedback from everyone that should be involved. If you\`re fixing a typo or
      something that\`s on fire 🔥 (e.g. incident related), you can skip this step.
      guide and passes our tests.

**A picture of a cute animal (not mandatory, but encouraged)**

![A hamster washing their paws in a miniature sink](https://media.giphy.com/media/Jq80I3ZiZh77sqhsUV/giphy.gif)
